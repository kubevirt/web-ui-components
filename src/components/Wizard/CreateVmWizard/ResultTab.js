import React from 'react';
import PropTypes from 'prop-types';

import { Loading } from '../../Loading';

const getCreationText = (state, createTemplate) => `Creation of VM ${createTemplate ? 'Template ' : ''}${state}`;

export const ResultTab = ({ isSuccessful, isCreateTemplate, children }) => {
  if (isSuccessful == null) {
    return <Loading key="progress" text={getCreationText('in progress', isCreateTemplate)} />;
  }
  const state = isSuccessful ? 'was successful' : 'failed';
  const icon = isSuccessful ? 'pficon-ok' : 'pficon-error-circle-o';

  return (
    <div className="wizard-pf-complete blank-slate-pf" key="success">
      <div className="wizard-pf-success-icon">
        <span className={`pficon ${icon}`} />
      </div>
      <h3 className="blank-slate-pf-main-action">{getCreationText(state, isCreateTemplate)}</h3>
      {children}
    </div>
  );
};

ResultTab.defaultProps = {
  children: null,
  isSuccessful: null,
  isCreateTemplate: false,
};

ResultTab.propTypes = {
  children: PropTypes.node,
  isSuccessful: PropTypes.bool,
  isCreateTemplate: PropTypes.bool,
};
