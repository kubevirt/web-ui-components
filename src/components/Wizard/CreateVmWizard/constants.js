export const BASIC_SETTINGS_TAB_IDX = 0;
export const NETWORK_TAB_IDX = 1;
export const STORAGE_TAB_IDX = 2;
export const RESULTS_TAB_IDX = 3;

export const ALL_TABS = [BASIC_SETTINGS_TAB_IDX, NETWORK_TAB_IDX, STORAGE_TAB_IDX, RESULTS_TAB_IDX];

// BasicSettingsTab
export const NAME_KEY = 'name';
export const NAMESPACE_KEY = 'namespace';
export const DESCRIPTION_KEY = 'description';
export const IMAGE_SOURCE_TYPE_KEY = 'imageSourceType';
export const REGISTRY_IMAGE_KEY = 'registryImage';
export const IMAGE_URL_KEY = 'imageURL';
export const USER_TEMPLATE_KEY = 'userTemplate';
export const OPERATING_SYSTEM_KEY = 'operatingSystem';
export const FLAVOR_KEY = 'flavor';
export const MEMORY_KEY = 'memory';
export const CPU_KEY = 'cpu';
export const WORKLOAD_PROFILE_KEY = 'workloadProfile';
export const START_VM_KEY = 'startVM';
export const CREATE_TEMPLATE_KEY = 'createTemplate ';
export const CLOUD_INIT_KEY = 'cloudInit';
export const HOST_NAME_KEY = 'hostname';
export const AUTHKEYS_KEY = 'authKeys';
