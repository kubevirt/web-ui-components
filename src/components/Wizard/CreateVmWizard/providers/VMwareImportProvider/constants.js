export const PROVIDER_VMWARE = 'VMware';

export const PROVIDER_VMWARE_VCENTER_KEY = 'vCenterInstance';
export const PROVIDER_VMWARE_HOSTNAME_KEY = 'vmwareHostname';
export const PROVIDER_VMWARE_USER_NAME_KEY = 'vmwareUserName';
export const PROVIDER_VMWARE_USER_PASSWORD_KEY = 'vmwarePassword';
export const PROVIDER_VMWARE_REMEMBER_PASSWORD_KEY = 'rememberVMwareCredentials';

export const PROVIDER_VMWARE_CHECK_CONNECTION_KEY = 'checkConnectionButton';
export const PROVIDER_VMWARE_CHECK_CONNECTION_BTN_TEXT_KEY = 'checkConnectionButtonText';
export const PROVIDER_VMWARE_CHECK_CONNECTION_BTN_SAVE = 'Check and Save';
export const PROVIDER_VMWARE_CHECK_CONNECTION_BTN_DONT_SAVE = 'Check';
export const PROVIDER_VMWARE_STATUS_KEY = 'vmwareStatus';

export const PROVIDER_VMWARE_VM_KEY = 'vmwareVm';

export const PROVIDER_VMWARE_V2V_NAME_KEY = 'v2vVmwareName';
export const PROVIDER_VMWARE_NEW_VCENTER_NAME_KEY = 'newVCenterName';

export const PROVIDER_VMWARE_V2V_LAST_ERROR = 'PROVIDER_VMWARE_V2V_LAST_ERROR';
