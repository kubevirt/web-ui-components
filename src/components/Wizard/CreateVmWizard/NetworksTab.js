import React from 'react';
import PropTypes from 'prop-types';
import { FormFactory } from '../../Form';
import { TableFactory } from '../../Table/TableFactory';
import { ACTIONS_TYPE, DELETE_ACTION } from '../../Table/constants';
import { POD_NETWORK } from '../../../constants';

const validateNetwork = network => {
  const errors = Array(4).fill(null);

  if (!network || network.id == null) {
    errors[0] = 'Empty entity.'; // row error on index 0
  }

  if (!network.name) {
    errors[1] = 'Name is empty';
  }

  if (network.id !== 1 && !network.network) {
    errors[3] = 'Network config must be selected';
  }

  return errors;
};

export const validateNetworksNamespace = (networkConfigs, namespace, networks) => {
  const availableNetworkConfigs = networkConfigs.filter(nc => nc.metadata.namespace === namespace);
  networks.filter(network => isBootableNetwork(network)).forEach(network => {
    if (!network.errors) {
      network.errors = Array(4).fill(null);
    }
    network.errors[3] = availableNetworkConfigs.some(nc => nc.metadata.name === network.network)
      ? null
      : 'Network config not found';
  });
};

export const hasError = network => (network.errors ? network.errors.some(error => !!error) : false);

export const isBootableNetwork = network => network.network !== POD_NETWORK && network.network !== '';

export class NetworksTab extends React.Component {
  constructor(props) {
    super(props);
    const rows = props.networks.map(({ id, name, mac, network, isBootable, errors }) => ({
      id,
      name,
      mac,
      network,
      errors,
      isBootable,
      renderConfig: 0,
      edit: false,
      editable: true
    }));

    validateNetworksNamespace(props.networkConfigs, props.namespace, rows);
    this.resolveBootableNetwork(props.pxeBoot, rows);
    this.publishResults(rows);
    this.state = {
      // eslint-disable-next-line
      nextId: rows.length + 1,
      editing: false,
      rows
    };
  }

  checkPxeBootable = rows => (this.props.pxeBoot ? rows.some(row => row.isBootable) : true);

  publishResults = rows => {
    let valid = this.checkPxeBootable(rows);
    const nics = rows.map(({ id, isBootable, name, mac, network, errors }) => {
      const result = {
        id,
        isBootable,
        name,
        mac,
        network,
        errors
      };

      if (valid && errors) {
        for (const error of errors) {
          if (error) {
            valid = false;
            break;
          }
        }
      }
      return result;
    });

    const value = {
      networks: nics
    };
    this.props.onChange(value, valid);
  };

  onRowActivate = rows => {
    this.setState({ rows, editing: true });
  };

  onRowUpdate = (rows, updatedRowId, editing) => {
    const updatedRow = rows.find(r => r.id === updatedRowId);
    if (updatedRow.isBootable && updatedRow.network === POD_NETWORK) {
      updatedRow.isBootable = false;
    }
    updatedRow.errors = validateNetwork(updatedRow);
    this.rowsChanged(rows, editing);
  };

  rowsChanged = (rows, editing) => {
    this.resolveBootableNetwork(this.props.pxeBoot, rows);
    this.publishResults(rows);
    this.setState({ rows, editing });
  };

  resolveBootableNetwork = (pxeBoot, rows) => {
    if (pxeBoot && !rows.some(row => row.isBootable && !hasError(row))) {
      const bootableNetworks = rows.filter(row => isBootableNetwork(row));
      if (bootableNetworks.length > 0) {
        let bootableFound = false;
        bootableNetworks.forEach(network => {
          if (!bootableFound && !hasError(network)) {
            network.isBootable = true;
            bootableFound = true;
          } else {
            network.isBootable = false;
          }
        });
      }
    }
  };

  createNic = () => {
    this.setState(state => ({
      nextId: state.nextId + 1,
      rows: [
        ...state.rows,
        {
          id: state.nextId,
          isBootable: false,
          editable: true,
          edit: true, // trigger immediate edit,
          name: `eth${state.nextId - 1}`,
          mac: '',
          network: '',
          renderConfig: 0
        }
      ]
    }));
  };

  getColumns = () => [
    {
      header: {
        label: 'NIC Name',
        props: {
          style: {
            width: '32%'
          }
        }
      },
      property: 'name',
      renderConfigs: [
        {
          id: 'name-edit',
          type: 'text'
        }
      ]
    },
    {
      header: {
        label: 'MAC Address',
        props: {
          style: {
            width: '32%'
          }
        }
      },
      property: 'mac',
      renderConfigs: [
        {
          id: 'mac-edit',
          type: 'text'
        }
      ]
    },
    {
      header: {
        label: 'Network Configuration',
        props: {
          style: {
            width: '32%'
          }
        }
      },
      property: 'network',
      renderConfigs: [
        {
          id: 'network-edit',
          type: 'dropdown',
          choices: this.props.networkConfigs
            .filter(networkConfig => networkConfig.metadata.namespace === this.props.namespace)
            .map(networkConfig => networkConfig.metadata.name)
            .concat(POD_NETWORK)
            .filter(networkConfig => !this.state.rows.some(r => r.network === networkConfig)),
          initialValue: '--- Select Network Definition ---'
        }
      ]
    },
    {
      header: {
        props: {
          style: {
            width: '4%'
          }
        }
      },
      renderConfigs: [
        {
          id: 'actions',
          type: ACTIONS_TYPE,
          actions: [
            {
              actionType: DELETE_ACTION,
              text: 'Remove NIC'
            }
          ],
          visibleOnEdit: false
        }
      ]
    }
  ];

  getActionButtons = () => [
    {
      className: 'create-network',
      onClick: this.createNic,
      id: 'create-network-btn',
      text: 'Create NIC',
      disabled: this.state.editing
    }
  ];

  getFormFields = pxeNetworks => ({
    pxeNetwork: {
      id: 'pxe-nic-dropdown',
      title: 'PXE NIC',
      type: 'dropdown',
      defaultValue: '--- Select PXE NIC ---',
      choices: pxeNetworks.map(n => n.name),
      required: true,
      help: 'Pod network is not PXE bootable'
    }
  });

  onFormChange = newValue => {
    this.setState(state => {
      state.rows.forEach(row => {
        row.isBootable = row.name === newValue;
      });
      this.publishResults(state.rows);
      return state.rows;
    });
  };

  render() {
    const columns = this.getColumns();
    const actionButtons = this.getActionButtons();

    let pxeForm;
    if (this.props.pxeBoot) {
      const pxeNetworks = this.state.rows.filter(row => isBootableNetwork(row) && !hasError(row));
      const bootableNetwork = pxeNetworks.find(row => row.isBootable);
      const values = {
        pxeNetwork: {
          value: bootableNetwork ? bootableNetwork.name : undefined,
          validMsg: pxeNetworks.length === 0 ? 'A PXE-capable NIC could not be found' : undefined
        }
      };

      pxeForm = (
        <FormFactory
          fields={this.getFormFields(pxeNetworks)}
          fieldsValues={values}
          onFormChange={this.onFormChange}
          textPosition="text-left"
          labelSize={2}
          controlSize={10}
          formClassName="pxe-form"
        />
      );
    }
    return (
      <React.Fragment>
        <TableFactory
          actionButtons={actionButtons}
          columns={columns}
          rows={this.state.rows}
          onRowUpdate={this.onRowUpdate}
          onRowDeleteOrMove={this.rowsChanged}
          onRowActivate={this.onRowActivate}
        />
        {pxeForm}
      </React.Fragment>
    );
  }
}

NetworksTab.propTypes = {
  onChange: PropTypes.func.isRequired,
  networks: PropTypes.array.isRequired,
  pxeBoot: PropTypes.bool.isRequired,
  networkConfigs: PropTypes.array.isRequired,
  namespace: PropTypes.string.isRequired
};
