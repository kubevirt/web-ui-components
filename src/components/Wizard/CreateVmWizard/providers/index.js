import { get } from 'lodash';

import { NAMESPACE_KEY, PROVIDER_KEY, PROVIDER_VMWARE } from '../constants';
import { isImageSourceType, settingsValue } from '../../../../k8s/selectors';
import { PROVISION_SOURCE_IMPORT } from '../../../../constants';

import { getVMWareSection } from './vmware';
import { startV2VVMWareController } from './v2vvmwareController';
import { HELP_PROVIDER_VMWARE } from '../strings';

const providerList = [PROVIDER_VMWARE];

const getProviderHelp = basicSettings => {
  const provider = settingsValue(basicSettings, PROVIDER_KEY);
  switch (provider) {
    case PROVIDER_VMWARE:
      return HELP_PROVIDER_VMWARE;
    default:
      return null;
  }
};

export const importProviders = (basicSettings, WithResources, k8sCreate, k8sGet, k8sPatch) => ({
  [PROVIDER_KEY]: {
    id: 'provider-dropdown',
    title: 'Provider',
    type: 'dropdown',
    defaultValue: '--- Select Provider ---',
    choices: providerList,
    required: true,
    onChange: (...opts) => onProviderChanged(k8sCreate, k8sGet, ...opts),
    isVisible: basicVmSettings => isImageSourceType(basicVmSettings, PROVISION_SOURCE_IMPORT),
    help: getProviderHelp(basicSettings),
  },
  ...getVMWareSection(basicSettings, WithResources, k8sCreate, k8sGet, k8sPatch),
});

const onProviderChanged = (k8sCreate, k8sGet, valueValidationPair, key, formValid, prevBasicSettings) => {
  startV2VVMWareController({ k8sCreate, k8sGet, namespace: get(prevBasicSettings[NAMESPACE_KEY], 'value') });
};
