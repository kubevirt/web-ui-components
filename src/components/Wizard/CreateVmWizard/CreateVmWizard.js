import React from 'react';
import PropTypes from 'prop-types';
import { findIndex } from 'lodash';
import { Wizard } from 'patternfly-react';

import BasicSettingsTab from './BasicSettingsTab';
import StorageTab from './StorageTab';
import ResultTab from './ResultTab';

import { createVm, createVmTemplate } from '../../../k8s/request';
import {
  POD_NETWORK,
  PROVISION_SOURCE_URL,
  PROVISION_SOURCE_CONTAINER,
  PROVISION_SOURCE_PXE,
} from '../../../constants';
import { NetworksTab } from './NetworksTab';
import { settingsValue } from '../../../k8s/selectors';
import {
  PROVISION_SOURCE_TYPE_KEY,
  USER_TEMPLATE_KEY,
  BASIC_SETTINGS_TAB_KEY,
  NETWORKS_TAB_KEY,
  STORAGE_TAB_KEY,
  RESULT_TAB_KEY,
  STORAGE_TYPE_CONTAINER,
  STORAGE_TYPE_DATAVOLUME,
  NETWORK_TYPE_POD,
} from './constants';
import {
  CREATE_VM,
  CREATE_VM_TEMPLATE,
  STEP_BASIC_SETTINGS,
  STEP_NETWORK,
  STEP_STORAGE,
  STEP_RESULT,
  NEXT,
} from './strings';
import { loadingWizardTab } from '../loadingWizardTab';
import {
  getUserTemplate,
  getTemplateStorages,
  getTemplateInterfaces,
  hasAutoAttachPodInterface,
} from '../../../utils/templates';

const LoadingBasicWizardTab = loadingWizardTab(BasicSettingsTab);
const LoadingStorageTab = loadingWizardTab(StorageTab);
const LoadingNetworksTab = loadingWizardTab(NetworksTab);

const getBasicSettingsValue = (stepData, key) => settingsValue(stepData[BASIC_SETTINGS_TAB_KEY].value, key);

const onUserTemplateChangedInStorageTab = (stepData, newUserTemplate) => {
  const withoutDiscardedTemplateStorage = stepData[STORAGE_TAB_KEY].value.filter(storage => !storage.templateStorage);

  const rows = [...withoutDiscardedTemplateStorage];

  if (newUserTemplate) {
    const templateStorages = getTemplateStorages(newUserTemplate).map(storage => ({
      templateStorage: storage,
    }));
    rows.push(...templateStorages);
  }

  return {
    ...stepData,
    [STORAGE_TAB_KEY]: {
      ...stepData[STORAGE_TAB_KEY],
      value: rows,
    },
  };
};

const onUserTemplateChangedInNetworksTab = (stepData, newUserTemplate) => {
  const withoutDiscardedTemplateNetworks = stepData[NETWORKS_TAB_KEY].value.filter(network => !network.templateNetwork);

  const rows = [...withoutDiscardedTemplateNetworks];
  if (!rows.find(row => row.rootNetwork)) {
    rows.push(podNetwork);
  }

  if (newUserTemplate) {
    const templateInterfaces = getTemplateInterfaces(newUserTemplate);
    const networks = templateInterfaces.map(i => ({
      templateNetwork: i,
    }));

    // do not add root interface if there already is one or autoAttachPodInterface is set to false
    if (networks.some(network => network.templateNetwork.network.pod) || !hasAutoAttachPodInterface(newUserTemplate)) {
      const index = findIndex(rows, network => network.rootNetwork);
      rows.splice(index, 1);
    }

    rows.push(...networks);
  }

  return {
    ...stepData,
    [NETWORKS_TAB_KEY]: {
      ...stepData[NETWORKS_TAB_KEY],
      value: rows,
    },
  };
};

const rootDisk = {
  rootStorage: {},
  name: 'rootdisk',
  isBootable: true,
};

export const rootContainerDisk = {
  ...rootDisk,
  storageType: STORAGE_TYPE_CONTAINER,
};

export const rootDataVolumeDisk = {
  ...rootDisk,
  storageType: STORAGE_TYPE_DATAVOLUME,
  size: 10,
};

const onImageSourceTypeChangedInStorageTab = stepData => {
  const withoutRootStorage = stepData[STORAGE_TAB_KEY].value.filter(storage => !storage.rootStorage);
  const rows = [...withoutRootStorage];
  const basicSettings = stepData[BASIC_SETTINGS_TAB_KEY].value;
  if (!(basicSettings[USER_TEMPLATE_KEY] && basicSettings[USER_TEMPLATE_KEY].value)) {
    let storage;
    const provisionSource = basicSettings[PROVISION_SOURCE_TYPE_KEY].value;
    switch (provisionSource) {
      case PROVISION_SOURCE_URL:
        storage = rootDataVolumeDisk;
        break;
      case PROVISION_SOURCE_CONTAINER:
        storage = rootContainerDisk;
        break;
      case PROVISION_SOURCE_PXE:
        break;
      default:
        // eslint-disable-next-line
        console.warn(`Unknown provision source ${provisionSource}`);
        break;
    }
    if (storage) {
      rows.push(storage);
    }
  }
  return {
    ...stepData,
    [STORAGE_TAB_KEY]: {
      ...stepData[STORAGE_TAB_KEY],
      value: rows,
    },
  };
};

const onUserTemplateChanged = (props, stepData) => {
  const userTemplateName = getBasicSettingsValue(stepData, USER_TEMPLATE_KEY);
  const userTemplate = userTemplateName ? getUserTemplate(props.templates, userTemplateName) : undefined;
  const newStepData = onUserTemplateChangedInStorageTab(stepData, userTemplate);
  return onUserTemplateChangedInNetworksTab(newStepData, userTemplate);
};

const onImageSourceTypeChanged = (props, stepData) => onImageSourceTypeChangedInStorageTab(stepData);

const podNetwork = {
  rootNetwork: {},
  id: 0,
  name: 'eth0',
  mac: '',
  network: POD_NETWORK,
  editable: true,
  edit: false,
  networkType: NETWORK_TYPE_POD,
};

export class CreateVmWizard extends React.Component {
  state = {
    activeStepIndex: 0,
    stepData: {
      [BASIC_SETTINGS_TAB_KEY]: {
        // Basic Settings
        value: {},
        valid: false,
      },
      [NETWORKS_TAB_KEY]: {
        value: [podNetwork],
        valid: true,
      },
      [STORAGE_TAB_KEY]: {
        value: [], // Storages
        valid: true, // empty Storages are valid
      },
      [RESULT_TAB_KEY]: {
        value: '',
        valid: null, // result of the request
      },
    },
  };

  getLastStepIndex = () => this.wizardStepsNewVM.length - 1;

  lastStepReached = () => this.state.activeStepIndex === this.getLastStepIndex();

  callbacks = [
    {
      field: USER_TEMPLATE_KEY,
      callback: onUserTemplateChanged,
    },
    {
      field: PROVISION_SOURCE_TYPE_KEY,
      callback: onImageSourceTypeChanged,
    },
  ];

  onStepDataChanged = (key, value, valid) => {
    this.setState((state, props) => {
      let stepData = { ...state.stepData };

      stepData[key] = {
        value,
        valid,
      };

      if (key === BASIC_SETTINGS_TAB_KEY) {
        this.callbacks.forEach(callback => {
          const oldValue = getBasicSettingsValue(state.stepData, callback.field);
          const newValue = getBasicSettingsValue(stepData, callback.field);
          if (oldValue !== newValue) {
            stepData = callback.callback(props, stepData);
          }
        });
      }

      return { stepData };
    });
  };

  finish() {
    const create = this.props.createTemplate
      ? createVmTemplate
      : (k8sCreate, templates, basicSettings, networks, storage) =>
          createVm(k8sCreate, templates, basicSettings, networks, storage, this.props.persistentVolumeClaims);
    create(
      this.props.k8sCreate,
      this.props.templates,
      this.state.stepData[BASIC_SETTINGS_TAB_KEY].value,
      this.state.stepData[NETWORKS_TAB_KEY].value,
      this.state.stepData[STORAGE_TAB_KEY].value
    )
      .then(results => {
        const vmResult = results[results.length - 1];
        return this.onStepDataChanged(RESULT_TAB_KEY, JSON.stringify(vmResult, null, 1), true);
      })
      .catch(error => this.onStepDataChanged(RESULT_TAB_KEY, error.message, false));
  }

  onStepChanged = newActiveStepIndex => {
    if (
      !this.lastStepReached() && // do not allow going back once last step is reached
      (newActiveStepIndex < this.state.activeStepIndex || // allow going back to past steps
        this.state.stepData[this.wizardStepsNewVM[newActiveStepIndex - 1].key].valid)
    ) {
      if (newActiveStepIndex === this.getLastStepIndex()) {
        this.finish();
      }
      this.setState({ activeStepIndex: newActiveStepIndex });
    }
  };

  wizardStepsNewVM = [
    {
      title: STEP_BASIC_SETTINGS,
      key: BASIC_SETTINGS_TAB_KEY,
      render: () => {
        const loadingData = {
          namespaces: this.props.namespaces,
          templates: this.props.templates,
        };
        return (
          <LoadingBasicWizardTab
            key="1"
            selectedNamespace={this.props.selectedNamespace}
            basicSettings={this.state.stepData[BASIC_SETTINGS_TAB_KEY].value}
            onChange={(value, valid) => this.onStepDataChanged(BASIC_SETTINGS_TAB_KEY, value, valid)}
            loadingData={loadingData}
            createTemplate={this.props.createTemplate}
          />
        );
      },
    },
    {
      title: STEP_NETWORK,
      key: NETWORKS_TAB_KEY,
      render: () => {
        const loadingData = {
          networkConfigs: this.props.networkConfigs,
        };
        const sourceType = getBasicSettingsValue(this.state.stepData, PROVISION_SOURCE_TYPE_KEY);
        return (
          <LoadingNetworksTab
            onChange={(value, valid) => this.onStepDataChanged(NETWORKS_TAB_KEY, value, valid)}
            networkConfigs={this.props.networkConfigs}
            networks={this.state.stepData[NETWORKS_TAB_KEY].value || []}
            sourceType={sourceType}
            namespace={this.state.stepData[BASIC_SETTINGS_TAB_KEY].value.namespace.value}
            loadingData={loadingData}
          />
        );
      },
    },
    {
      title: STEP_STORAGE,
      key: STORAGE_TAB_KEY,
      render: () => {
        const loadingData = {
          storageClasses: this.props.storageClasses,
          persistentVolumeClaims: this.props.persistentVolumeClaims,
        };
        const sourceType = getBasicSettingsValue(this.state.stepData, PROVISION_SOURCE_TYPE_KEY);
        return (
          <LoadingStorageTab
            initialStorages={this.state.stepData[STORAGE_TAB_KEY].value}
            onChange={(value, valid) => this.onStepDataChanged(STORAGE_TAB_KEY, value, valid)}
            units={this.props.units}
            sourceType={sourceType}
            namespace={this.state.stepData[BASIC_SETTINGS_TAB_KEY].value.namespace.value}
            loadingData={loadingData}
          />
        );
      },
    },
    {
      title: STEP_RESULT,
      key: RESULT_TAB_KEY,
      render: () => {
        const stepData = this.state.stepData[RESULT_TAB_KEY];
        return (
          <ResultTab result={stepData.value} success={stepData.valid} createTemplate={this.props.createTemplate} />
        );
      },
    },
  ];

  render() {
    const beforeLastStepReached = this.state.activeStepIndex === this.wizardStepsNewVM.length - 2;
    const lastStepReached = this.lastStepReached();

    const createVmText = this.props.createTemplate ? CREATE_VM_TEMPLATE : CREATE_VM;

    return (
      <Wizard.Pattern
        show
        onHide={this.props.onHide}
        steps={this.wizardStepsNewVM}
        activeStepIndex={this.state.activeStepIndex}
        onStepChanged={index => this.onStepChanged(index)}
        previousStepDisabled={lastStepReached}
        cancelButtonDisabled={lastStepReached}
        stepButtonsDisabled={lastStepReached}
        nextStepDisabled={!this.state.stepData[this.wizardStepsNewVM[this.state.activeStepIndex].key].valid}
        nextText={beforeLastStepReached ? createVmText : NEXT}
        title={createVmText}
      />
    );
  }
}

CreateVmWizard.defaultProps = {
  selectedNamespace: undefined,
  templates: undefined,
  namespaces: undefined,
  networkConfigs: undefined,
  persistentVolumeClaims: undefined,
  storageClasses: undefined,
  createTemplate: false,
};

CreateVmWizard.propTypes = {
  onHide: PropTypes.func.isRequired,
  templates: PropTypes.array,
  namespaces: PropTypes.array,
  selectedNamespace: PropTypes.object,
  k8sCreate: PropTypes.func.isRequired,
  networkConfigs: PropTypes.array,
  persistentVolumeClaims: PropTypes.array,
  storageClasses: PropTypes.array,
  units: PropTypes.object.isRequired,
  createTemplate: PropTypes.bool,
};
