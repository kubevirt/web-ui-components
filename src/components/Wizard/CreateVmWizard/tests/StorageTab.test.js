import React from 'react';
import { shallow, mount } from 'enzyme';
import StorageTab from '../StorageTab';

import { units, persistentVolumeClaims, storageClasses } from '../fixtures/CreateVmWizard.fixture';
import { rows } from '../fixtures/StorageTab.fixture';
import { PROVISION_SOURCE_URL } from '../../../../constants';

const testStorageTab = (onChange, initialDisks) => (
  <StorageTab
    persistentVolumeClaims={persistentVolumeClaims}
    storageClasses={storageClasses}
    onChange={onChange || jest.fn()}
    initialStorages={initialDisks || []}
    units={units}
    sourceType={PROVISION_SOURCE_URL}
    namespace="default"
  />
);

const thirdRow = rows[2];
const testRows = rows.slice(0, 2);

const updatedThirdRow = {
  ...thirdRow,
  attachStorage: {
    id: 'attach',
  },
  name: 'name',
};

describe('<StorageTab />', () => {
  it('renders correctly', () => {
    const component = shallow(testStorageTab());
    expect(component).toMatchSnapshot();
  });

  // TODO: add create disk test once the create blank disk feature is implemented in kubevirt
  // it('creates disk', () => {
  //   const component = shallow(testStorageTab());
  //
  //   expect(component.state().rows).toHaveLength(0);
  //   component.instance().createDisk();
  //
  //   expect(component.state().rows).toHaveLength(1);
  // });

  it('attaches storage', () => {
    const component = shallow(testStorageTab());

    expect(component.state().rows).toHaveLength(0);
    component.instance().attachStorage();

    expect(component.state().rows).toHaveLength(1);
  });

  it('initializes attach storage ', () => {
    const onChange = jest.fn();
    const component = mount(testStorageTab(onChange, testRows.slice(0, 1)));

    expect(component.state().rows).toHaveLength(1);
  });

  it('initializes disk and remove it', () => {
    const onChange = jest.fn();
    const component = mount(testStorageTab(onChange, testRows.slice(0, 1)));

    expect(component.state().rows).toHaveLength(1);
    component.instance().setState({ rows: [] });
    expect(component.state().rows).toHaveLength(0);
  });

  it('calls onChange when rows change', () => {
    const onChange = jest.fn();

    const component = mount(testStorageTab(onChange, [...testRows, thirdRow]));
    expect(onChange).toHaveBeenCalledTimes(1);

    component.instance().rowsChanged(testRows, false);

    expect(onChange).toHaveBeenCalledTimes(2);
  });

  it('calls onChange when row is updated', () => {
    const onChange = jest.fn();

    const newRows = [...testRows, thirdRow];
    const component = mount(testStorageTab(onChange, newRows));
    expect(onChange).toHaveBeenCalledTimes(1);

    component.instance().onRowUpdate(newRows, thirdRow.id, false);

    expect(onChange).toHaveBeenCalledTimes(2);
  });

  it('does not call onChange when row is activated', () => {
    const onChange = jest.fn();

    const component = mount(testStorageTab(onChange, [...testRows, thirdRow]));
    expect(onChange).toHaveBeenCalledTimes(1);

    component.instance().onRowActivate([...testRows, updatedThirdRow]);

    expect(onChange).toHaveBeenCalledTimes(1);
  });
});
