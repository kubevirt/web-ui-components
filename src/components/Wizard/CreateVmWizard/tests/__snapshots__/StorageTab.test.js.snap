// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<StorageTab /> renders correctly 1`] = `
<Fragment>
  <TableFactory
    actionButtons={
      Array [
        Object {
          "className": "kubevirt-create-vm-wizard__button-create-disk",
          "disabled": false,
          "id": "create-storage-btn",
          "onClick": [Function],
          "text": "Create Disk",
        },
        Object {
          "className": "kubevirt-create-vm-wizard__button-attach-disk",
          "disabled": false,
          "id": "attach-disk-btn",
          "onClick": [Function],
          "text": "Attach Disk",
        },
      ]
    }
    columns={
      Array [
        Object {
          "header": Object {
            "label": "Disk Name",
            "props": Object {
              "style": Object {
                "width": "50%",
              },
            },
          },
          "property": "name",
          "renderConfigs": Array [
            Object {
              "id": "name-edit",
              "type": "text",
            },
            Object {
              "choices": Array [
                "disk-one",
                "disk-three",
              ],
              "id": "name-attach-edit",
              "initialValue": "--- Select Storage ---",
              "type": "dropdown",
            },
            Object {
              "id": "name-edit",
              "type": "text",
            },
            Object {
              "id": "name-edit",
              "type": "text",
            },
            Object {
              "id": "name-edit",
              "type": "text",
            },
          ],
        },
        Object {
          "header": Object {
            "label": "Size (GB)",
            "props": Object {
              "style": Object {
                "width": "23%",
              },
            },
          },
          "property": "size",
          "renderConfigs": Array [
            Object {
              "id": "size-edit",
              "type": "positive-number",
            },
            undefined,
            undefined,
            undefined,
            Object {
              "id": "size-edit",
              "type": "positive-number",
            },
          ],
        },
        Object {
          "header": Object {
            "label": "Storage Class",
            "props": Object {
              "style": Object {
                "width": "23%",
              },
            },
          },
          "property": "storageClass",
          "renderConfigs": Array [
            Object {
              "choices": Array [
                "nfs",
                "iscsi",
                "glusterfs",
                "azuredisk",
              ],
              "id": "storage-edit",
              "initialValue": "--- Select ---",
              "type": "dropdown",
            },
            undefined,
            undefined,
            undefined,
            Object {
              "choices": Array [
                "nfs",
                "iscsi",
                "glusterfs",
                "azuredisk",
              ],
              "id": "storage-edit",
              "initialValue": "--- Select ---",
              "type": "dropdown",
            },
          ],
        },
        Object {
          "header": Object {
            "props": Object {
              "style": Object {
                "width": "4%",
              },
            },
          },
          "renderConfigs": Array [
            Object {
              "actions": Array [
                Object {
                  "actionType": "delete",
                  "text": "Remove Disk",
                },
              ],
              "id": "actions",
              "type": "actions",
              "visibleOnEdit": false,
            },
            Object {
              "actions": Array [
                Object {
                  "actionType": "delete",
                  "text": "Remove Disk",
                },
              ],
              "id": "actions",
              "type": "actions",
              "visibleOnEdit": false,
            },
            Object {
              "actions": Array [
                Object {
                  "actionType": "delete",
                  "text": "Remove Disk",
                },
              ],
              "id": "actions",
              "type": "actions",
              "visibleOnEdit": false,
            },
          ],
        },
      ]
    }
    onRowActivate={[Function]}
    onRowDeleteOrMove={[Function]}
    onRowUpdate={[Function]}
    rows={Array []}
  />
  <FormFactory
    controlSize={10}
    fields={
      Object {
        "bootableDisk": Object {
          "choices": Array [],
          "defaultValue": "--- Select Bootable Disk ---",
          "disabled": true,
          "id": "bootable-disk-dropdown",
          "required": true,
          "title": "Bootable Disk",
          "type": "dropdown",
        },
      }
    }
    fieldsValues={
      Object {
        "bootableDisk": Object {
          "validation": Object {
            "message": "A bootable disk could not be found",
            "type": "error",
          },
          "value": undefined,
        },
      }
    }
    formClassName="kubevirt-create-vm-wizard__pxe-form"
    labelSize={2}
    onFormChange={[Function]}
    textPosition="text-left"
  />
</Fragment>
`;
